File: Leetcode_Code/20. Valid Parentheses/20.cpp
class Solution {
public:
    bool isValid(string s) {
        map<char, char> dict = {{'[', ']'},
                                {'(', ')'},
                                {'{', '}'},
                                {'<', '>'}};
        stack<char> stk;
        for (int i = 0; i < s.length(); ++i) {
            if (dict.find(s[i]) != dict.end()) stk.push(dict[s[i]]);
            else {
                if (stk.size() == 0) return false;
                else {
                    if (stk.top() == s[i]) stk.pop();
                    else return false;
                }
            }
        }
        return stk.size() == 0;
    }
};

File: Leetcode_Code/20. Valid Parentheses/20.java
public class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> dict = new HashMap<>();
        dict.put('[', ']');
        dict.put('(', ')');
        dict.put('{', '}');
        dict.put('<', '>');
        Stack<Character> stk = new Stack<>();
        for (int i = 0; i < s.length(); ++i) {
            if (dict.containsKey(s.charAt(i))) stk.push(dict.get(s.charAt(i)));
            else {
                if (stk.isEmpty()) return false;
                else {
                    if (stk.peek() == s.charAt(i)) stk.pop();
                    else return false;
                }
            }
        }
        return stk.isEmpty();
    }
}
